File: leetcode/Binary_Tree/0226_Invert_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// by iteration
class Solution {
public:
	TreeNode* invertTree(TreeNode* root) {
		stack<TreeNode*> s;
		if (root != NULL)
			s.push(root);
		while (!s.empty()) {
			TreeNode *temp = s.top() -> left;
			s.top() -> left = s.top() -> right;
			s.top() -> right = temp;
			temp = s.top();
			s.pop();
			if (temp -> left != NULL)
				s.push(temp -> left);
			if (temp -> right != NULL)
				s.push(temp -> right);
		}

		return root;
	}
};

// by recursion
class Solution {
public:
	TreeNode* invertTree(TreeNode* root) {
		if (root == NULL)
			return root;
		TreeNode *temp = root -> left;
		root -> left = root -> right;
		root -> right = temp;
		
		invertTree(root -> left);
		invertTree(root -> right);
		return root;
	}
};

File: leetcode/Binary_Tree/0226_Invert_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# recursion
class Solution(object):
    def invertTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        if not root:
            return root

        root.left, root.right = root.right, root.left
        self.invertTree(root.left)
        self.invertTree(root.right)

        return root

# iteration
class Solution(object):
    def invertTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        if not root:
            return root

        stack = [root]
        while stack:
            current_node = stack.pop()
            current_node.left, current_node.right = current_node.right, current_node.left
            if current_node.left:
                stack.append(current_node.left)
            if current_node.right:
                stack.append(current_node.right)

        return root
